compiler-rt/lib/radsan/radsan.cpp

```diff
@@ -7,26 +7,137 @@
 */
 
 #include <radsan/radsan.h>
+
 #include <radsan/radsan_context.h>
+#include <radsan/radsan_flags.h>
 #include <radsan/radsan_interceptors.h>
-#include <unistd.h>
+
+#include "sanitizer_common/sanitizer_atomic.h"
+#include "sanitizer_common/sanitizer_common.h"
+#include "sanitizer_common/sanitizer_flag_parser.h"
+#include "sanitizer_common/sanitizer_flags.h"
+
+using namespace __sanitizer;
+
+namespace radsan {
+static __sanitizer::atomic_uint8_t radsan_inited{};
+static __sanitizer::atomic_uint8_t radsan_init_is_running{};
+static __sanitizer::atomic_uint64_t radsan_report_count{};
+static Flags radsan_flags{};
+
+Flags *flags() { return &radsan_flags; }
+
+void EnsureInitialized() {
+  CHECK(!radsan::IsInitRunning());
```

Comment:
Does this `return` if init is currently running?

Reply:
This is an `assert` basically, it will abort the process if this is happening.

This pattern was borrowed from msan, but I think the idea is: `EnsureInitialized` with MAKE SURE this thing is initialized before letting you continue. It is up to the caller to behave differently if the initialization is running.


You can see this check for IsInitRunning then the call here:

https://github.com/realtime-sanitizer/llvm-project/pull/13/files#diff-2d332ab3b2cfde868d5292ad94bfb45f304711efa30b0518c922ded7ff02eae7R39-R43


This is helpful for the case where initalization is running in one thread, and you have some interceptor that happens in another thread under conditions in which it would normally intercept. In that case it should just return the REAL functionality